### PR TITLE
Blrubenstein/not found page

### PR DESCRIPTION
**Pull request by blrubenstein**

### CODE REVIEW

File: ui/src/app/app-routing.module.ts [hunk 1]
```diff
@@ -52,6 +53,14 @@ const routes: Routes = [
             component: WorkspaceListComponent,
             data: {title: 'View Workspaces'}
           },
+          {
+            path: 'workspace/:ns/:wsid/notfound',
```
[the remark this answers is not in the document]
Comment (blrubenstein) [reply ui/src/app/app-routing.module.ts]: Yeah, this has the advantage of nice handling for workspaces/description.

File: ui/src/app/resolvers/workspace.ts [hunk 2]
```diff
@@ -24,8 +27,13 @@ export class WorkspaceResolver implements Resolve<WorkspaceData> {
 
     const call = this.api
       .getWorkspace(ns, wsid)
-      .map(({workspace, accessLevel}) => ({...workspace, accessLevel}));
-
+      .map(({workspace, accessLevel}) => ({...workspace, accessLevel}))
+      .catch(
+        (e) => {
+          this.router.navigate(['workspace', ns, wsid, 'notfound']);
```
[the remark this answers is not in the document]
Comment (blrubenstein) [reply ui/src/app/resolvers/workspace.ts]: Will do.

**Comment by blrubenstein** [comment]
Mohs also brought up that if you refresh the page on a not found page, and the workspace was created/shared during that time, it will still give a not found.

File: ui/src/app/app-routing.module.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (blrubenstein) [reply ui/src/app/app-routing.module.ts]: I am going to make it generic, but have it take in modes through the path, similar to WorkspaceEdit.

**Comment by johnstill** [comment]
> Mohs also brought up that if you refresh the page on a not found page, and the workspace was created/shared during that time, it will still give a not found.

Yeah, but thats not surprising - refreshing the page you're on should land you back at the page you're on.  If you want some magic redirection from the 404 page _back_ to the page that was originally requested you'll need to store the original url in the 404 page's url as a query param and do something with that.

On another topic - my recommendation is to have a generic 404 _route_ that you parametrize with a message directly.  I think that making your 404 page component explicitly aware of what kind of resource was requested, and attaching it to specialized 404 routes, is going to be much more maintenance for not much more value.  Again - this is my _recommendation_; I'm not going to block your PR on it.

**Comment by blrubenstein** [comment]
Let me think on that over lunch. I think it would be much better to have it redirect back to the resource (like, here it is routing back to the home page) but I think that the router actually maintains a history? 

However, we can't do that without knowing what resource it was requesting so we can query for it on refresh. Thinking through more seems valuable.

**Comment by blrubenstein** [comment]
Okay, this is the new pattern, per discussion with @calbach. PTAL when you get the chance.

**Comment by johnstill** [comment]
Listening for `NavigationError`s... I like it.

File: ui/src/app/resolvers/workspace.ts [hunk 3]
```diff
@@ -24,8 +27,11 @@ export class WorkspaceResolver implements Resolve<WorkspaceData> {
 
     const call = this.api
       .getWorkspace(ns, wsid)
-      .map(({workspace, accessLevel}) => ({...workspace, accessLevel}));
-
+      .map(({workspace, accessLevel}) => ({...workspace, accessLevel}))
+      .catch((e) => {
```
Comment (calbach) [inline ui/src/app/resolvers/workspace.ts]: This should verify that the code is actually a 404 first

File: ui/src/app/views/not-found/component.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import {NavigationError, Router} from '@angular/router';
+
+// Mostly borrowed from the Job Manager.
+@Component({
+    selector: 'app-not-found',
+    templateUrl: './component.html',
+    styleUrls: ['./component.css']
+})
+@Injectable()
+export class NotFoundComponent {
```
Comment (calbach) [inline ui/src/app/views/not-found/component.ts]: May want to name more generically. This is listening on any failed initial navigation regardless of the status code.

File: ui/src/app/views/not-found/component.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import {Component, Injectable} from '@angular/core';
+import {NavigationError, Router} from '@angular/router';
+
+// Mostly borrowed from the Job Manager.
+@Component({
+    selector: 'app-not-found',
+    templateUrl: './component.html',
+    styleUrls: ['./component.css']
+})
+@Injectable()
```
Comment (johnstill) [inline ui/src/app/views/not-found/component.ts]: I'm sorry, in my initial look the diff formatting kind of obscured the fact that the "service" and the "component" are the same thing.

I don't know how yet but I bet this leads to issues.  It seems like Angular's DI system is oddly permissive (I've also seen a component that was marked as a module, and there's that business where we are or were directly injecting AppComponent) but I don't see this behavior endorsed or documented anywhere and I'm willing to bet it works by accident and shouldn't be relied on.

Actually why does this component need `Injectable`? Won't it work the same without listing it as a provider?

**Comment by blrubenstein** [comment]
I will be doing a followup, because I realized I left some old code/lost some code that should be in here. Should have that ready soon.

**Pull request merged**